packages/docusaurus-plugin-search/src/index.ts

```diff
@@ -0,0 +1,84 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import {Plugin, Props, LoadContext} from '@docusaurus/types';
+import {promisify} from 'util';
+
+import scrap from './scrapper';
+
+import glob from 'glob';
+import {pluginOptions} from './types';
+import {readFile, writeFile} from 'fs-extra';
+import {join, dirname} from 'path';
+import Fuse from 'fuse.js';
+
+const flat = <T>(arr: Array<Array<T>>) =>
+  arr.reduce((acc, curr) => [...acc, ...curr]);
+
+export default function search(
+  _: LoadContext,
+  options: pluginOptions,
+): Plugin<unknown> {
+  console.log(options);
+  return {
+    name: 'docusaurus-plugin-search',
+
+    getThemePath() {
+      return join(__dirname, '..', 'theme');
+    },
+
+    getClientModules() {
+      const modules = [require.resolve('infima/dist/css/default/default.css')];
+
+      return modules;
+    },
+
+    getRootWrapper() {
+      return join(__dirname, '..', 'theme', 'wrapRoot.js');
+    },
+
+    async postBuild(props: Props) {
+      const files = await Promise.all(
+        options.include.map((pattern) =>
+          promisify(glob)(`${pattern}/index.html`, {
+            cwd: props.outDir,
+            ignore: options.exclude,
+          }),
+        ),
+      );
+      const uniquefiles = new Set(flat(files));
+      const scrapedData = await Promise.all(
+        Array.from(uniquefiles).map(async (file) => {
+          const content = await readFile(join(props.outDir, file));
+          return {url: dirname(file), dataNodes: scrap(String(content))};
+        }),
+      );
+      let id = 0;
+      const results = flat(
+        scrapedData.map((page) => {
+          return page.dataNodes.map((dataNode) => {
+            id += 1;
+            return {
+              pageID: page.url,
+              id,
+              ...dataNode,
+            };
+          });
+        }),
+      );
+      const index = Fuse.createIndex(['body'], results);
```

Comment:
You should rather not index body directly, but create multiple text fields with different scoring strategies. We need to be able to leverage the score from Fuse directly, and it's very important to have an index that is lightweight + efficient, instead of sorting the fuse result on each keystroke using a custom scoring algo.

Reply:
We don't index body directly, I wrote a copy of docsearch from algolia. I read their docs and implemented it in the node. I copied docsearch scrapper in nodejs. There are Four types H1,H2, H3, and p. Then I extract 20 best results and use them. I apply docs scoring algo on these results. It is the same as docsearch uses. Results from my plugin are close to results from algolia.